Frameworks/Foundation/NSProgress.mm

```diff
@@ -200,47 +200,77 @@ - (void)addChild:(NSProgress*)child withPendingUnitCount:(int64_t)inUnitCount {
+            // See NSProgress class reference:
+            // If you don’t create any child progress objects between the calls to becomeCurrentWithPendingUnitCount: and resignCurrent,
+            // the "parent" progress automatically updates its completedUnitCount by adding the pending units.
+            if (!currentProgress.childCreated) {
+                @synchronized(self) {
+                    [self setCompletedUnitCount:self.completedUnitCount + currentProgress.pendingUnitCountToAssign];
+                }
+            }
 
-    } else {
-        [NSException raise:NSInvalidArgumentException
-                    format:@"NSProgress was not the current progress on this thread %@", [NSThread currentThread]];
+            currentProgressStack->pop();
+            return;
+        }
     }
+
+    // self was nil, or otherwise was not currentProgress
+    [NSException raise:NSInvalidArgumentException
+                format:@"NSProgress was not the current progress on this thread %@", [NSThread currentThread]];
 }
 
 /**
  @Status Interoperable
 */
 - (void)setCompletedUnitCount:(int64_t)inUnitCount {
-    // Override synthesized setter so that parent can be updated with pendingUnitCount if necessary
+    // Override synthesized setter so that parent is updated as well
     @synchronized(self) { // Property is atomic
-        if ((_parent) && (inUnitCount >= _totalUnitCount) && (inUnitCount != _completedUnitCount)) {
-            [_parent _incrementCompletedUnitCount:_parentPendingUnitCount];
+        if (inUnitCount != _completedUnitCount) { // Only do anything if the count actually changed
+            int64_t deltaCompletedUnit = inUnitCount - _completedUnitCount;
+            double deltaFraction = static_cast<double>(deltaCompletedUnit) / _totalUnitCount;
+            [self _updateCompletedUnitsBy:deltaCompletedUnit fractionCompletedBy:deltaFraction unitCountForFraction:_totalUnitCount];
         }
+    }
+}
 
+- (void)_updateCompletedUnitsBy:(int64_t)deltaCompletedUnit
+            fractionCompletedBy:(double)deltaFraction
+           unitCountForFraction:(int64_t)unitCountForFraction {
+    @synchronized(self) {
         [self willChangeValueForKey:@"fractionCompleted"];
         [self willChangeValueForKey:@"indeterminate"];
-        _completedUnitCount = inUnitCount;
+        double prevFraction = _fractionCompleted;
+
+        // If this function is called from a child, deltaFraction needs to be adjusted according to the pending unit count of the child
+        // If this function is called from self, the latter half of the below expression evaluates to 1, and this is just deltaFraction
+        double deltaFractionForSelf = deltaFraction * unitCountForFraction / _totalUnitCount;
+
+        _completedUnitCount += deltaCompletedUnit;
+        _fractionCompleted += deltaFractionForSelf;
+
+        if (_fractionCompleted < 0) {
+            _fractionCompleted = 0;
+        }
         [self didChangeValueForKey:@"indeterminate"];
```

Comment:
since we're manually setting `_completedUnitCount`, we must trigger a `willChange` and `didChange` for it.